### PR TITLE
Add length-normalized DPO

### PR DESCRIPTION
**Pull request by hugoabonizio**
# What does this PR do?

This PR adds the length-normalized DPO used in Tulu-3 and introduced in the SimPO paper under `dpo_norm` loss type. This implementation and naming were inspired by the Open-Instruct implementation [[1]](https://github.com/allenai/open-instruct/blob/26b7325903813dd05ed7f4f04fa49f2794ef3b97/open_instruct/dpo_tune.py#L956) [[2]](https://github.com/allenai/open-instruct/blob/26b7325903813dd05ed7f4f04fa49f2794ef3b97/open_instruct/dpo_utils.py#L167-L168).

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by qgallouedec** [comment]
Hey, thanks for contributing!

Is it really a loss type? It seems to me that it can be combined with any loss type, no? 

What about having a new arg in `DPOConfig`? maybe `length_normalize`?

Also, I'd add a test for this

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -1170,6 +1170,11 @@ def concatenated_forward(self, model: nn.Module, batch: dict[str, Union[list, to
         per_token_logps[~loss_mask] = 0
         all_logps = per_token_logps.sum(-1)
 
+        if self.loss_type == "dpo_norm":
+            all_logps = (per_token_logps * loss_mask).sum(-1) / loss_mask.sum(-1)
+        else:
+            all_logps = per_token_logps.sum(-1)
+
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:1177]: `per_token_logps` is already 0 where the loss mask is 0. See two lines above. So I guess here it's equivalent to

```suggestion
        if self.loss_type == "dpo_norm":
            all_logps /= loss_mask.sum(-1)
```

right?

**Comment by qgallouedec** [comment]
No recent activity on this branch for few months, I'm closing this PR. Please feel free to reopen a PR if there is new activity.
